app/controllers/firebase_topics_controller.ts

```diff
@@ -0,0 +1,138 @@
+import vine from "@vinejs/vine";
+import { DateTime } from "luxon";
+
+import type { HttpContext } from "@adonisjs/core/http";
+import router from "@adonisjs/core/services/router";
+import { Constructor, LazyImport } from "@adonisjs/core/types/http";
+
+import {
+  CreateHookContext,
+  PartialModel,
+  RouteConfigurationOptions,
+} from "#controllers/base_controller";
+import { BadRequestException } from "#exceptions/http_exceptions";
+import FirebaseTopic, { TOPIC_NAME_REGEX } from "#models/firebase_topic";
+import PushNotificationService from "#services/push_notification_service";
+
+const { default: BaseController } = await (() =>
+  import("#controllers/base_controller"))();
+
+const topicToggleParamSchema = vine.object({
+  activate: vine.boolean(),
+});
+
+const topicStateParamSchema = vine.object({
+  deactivatedSince: vine.luxonDateTime().before(DateTime.now()),
+});
+
+const pushDataValidator = vine.compile(
+  vine.object({
+    notification: vine.object({
+      title: vine.string(),
+      body: vine.string(),
+      data: vine.record(vine.string()),
+    }),
+    topics: vine.array(vine.string()).notEmpty(),
+  }),
+);
+
+export default class FirebaseTopicController extends BaseController<
+  typeof FirebaseTopic
+> {
+  $configureRoutes(
+    controller: LazyImport<Constructor<FirebaseTopicController>>,
+    _?: RouteConfigurationOptions,
+  ) {
+    super.$configureRoutes(controller, {
+      skipRoutes: {
+        index: true,
+        destroy: true,
+      },
+    });
+    router
+      .patch("/topic/toggle/:id", [controller, "toggleTopicState"])
+      .as("toggle");
+    router.get("/topic/state", [controller, "indexTopicState"]).as("state");
+    router
+      .post("/notification/broadcast", [
+        controller,
+        "broadcastPushNotification",
+      ])
+      .as("broadcast");
+  }
```

Comment:
the base controller methods here will not have `/topic` prefixes, while other methods will, which may be confusing.
i'd rename this controller to just `firebase_controller.ts`, and wrap the call to the base controller's configure route method in `router.group` with the `/topic` prefix.